src/operations/extension/transpose.hpp

```diff
@@ -0,0 +1,221 @@
+SYCL_BLAS_INLINE void
+Transpose<in_place, Tile_size, wg_size, cl_size, local_memory, in_t, out_t,
+          element_t>::bind(cl::sycl::handler &cgh) {
+  A_.bind(cgh);
+  At_.bind(cgh);
+}
+
+template <bool in_place, int Tile_size, int wg_size, int cl_size,
+          bool local_memory, typename in_t, typename out_t, typename element_t>
+SYCL_BLAS_INLINE typename in_t::index_t
+Transpose<in_place, Tile_size, wg_size, cl_size, local_memory, in_t, out_t,
+          element_t>::get_size() const {
+  // Smallest TileSize square-multiple containing input/output matrices
+  return (M_pad_ * N_pad_);
+}
+
+template <bool in_place, int Tile_size, int wg_size, int cl_size,
+          bool local_memory, typename in_t, typename out_t, typename element_t>
+SYCL_BLAS_INLINE void
+Transpose<in_place, Tile_size, wg_size, cl_size, local_memory, in_t, out_t,
+          element_t>::adjust_access_displacement() {
+  A_.adjust_access_displacement();
+  At_.adjust_access_displacement();
+}
+
+/*!
+ *@brief get_indices. This function is used in the non-local memory kernel to
+ *compute global input & output indices.
+ *
+ * @param id [input] the cl::sycl::nd_item<1> of the current work_item
+ * @param in_idx [output] the input global-memory index
+ * @param out_idx [output] the output global-memory index
+ * @param i [output] the global row-index
+ * @param j [output] the global col-index
+ */
+template <bool in_place, int Tile_size, int wg_size, int cl_size,
+          bool local_memory, typename in_t, typename out_t, typename element_t>
+SYCL_BLAS_INLINE void
+Transpose<in_place, Tile_size, wg_size, cl_size, local_memory, in_t, out_t,
+          element_t>::get_indices(cl::sycl::nd_item<1> id, index_t &in_idx,
+                                  index_t &out_idx, index_t &i, index_t &j) {
+  index_t idg = id.get_group(0);
+  index_t idc = id.get_local_id(0);
+
+  const index_t jg = idg / tile_count_m_;
+  const index_t ig = idg - jg * tile_count_m_;
+
+  const index_t jl = idc / Tile_size;
+  const index_t il = idc - jl * Tile_size;
+
+  const index_t i_block_start = ig * Tile_size;
+  const index_t j_block_start = jg * Tile_size;
+
+  i = (i_block_start + il) * inc_a_;
+  j = (j_block_start + jl) * inc_at_;
+
+  in_idx =
+      i_block_start * inc_a_ + j_block_start * lda_ + il * inc_a_ + jl * lda_;
+
+  out_idx = i_block_start * ldat_ + j_block_start * inc_at_ + jl * inc_at_ +
+            il * ldat_;
+}
+
+template <bool in_place, int Tile_size, int wg_size, int cl_size,
+          bool local_memory, typename in_t, typename out_t, typename element_t>
+SYCL_BLAS_INLINE void
+Transpose<in_place, Tile_size, wg_size, cl_size, local_memory, in_t, out_t,
+          element_t>::eval(cl::sycl::nd_item<1> id) {
+  index_t idx = id.get_global_linear_id();
+
+  index_t in_index, out_index, i_id, j_id;
+
+  get_indices(id, in_index, out_index, i_id, j_id);
+
+  if (i_id < ((M_ - 1) * inc_a_ + 1)) {
+    auto A = A_.get_data().get_pointer();
```

Comment:
```suggestion
    auto A = A_.get_pointer();
```
This should work for both USM and buffer.

*Apply this to all `.get_data()` occurrences.*